8303002: Reject packed structs from linker

JornVernee opened the pull request:
This patch adds checks in AbstractLinker to reject packed structs and structs with excess padding (e.g. unnamed bitfields), since both of those are currently not supported, and the ABI/spec seems too vague to base support on.

/solves 8300784
/solves 8304803

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -195,6 +196,10 @@
  * </tbody>
  * </table></blockquote>
+ * <p>
+ * Note that due to limited ABI specification coverage, none of the native linker implementations supports
```
Is there something bigger to say here? E.g. can I pass a C int with an alignment other than its size? E.g. is there a requirement that the alignment of _all_ layouts should be identical to their natural alignment? If so, we should say so.

JornVernee replied:
I've added a list of constraints to the linker class javadoc which covers this.

JornVernee commented:
/solves 8303524

JornVernee commented:
Added the missing javadoc and also fixed an issue I found where the current check was not catching excess trailing padding that still resulted in the size of a group being a multiple of its alignment constraint (added a test case for that as well)

JornVernee commented:
/csr needed

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -195,6 +196,18 @@
  * </tbody>
  * </table></blockquote>
+ * <p>
+ * Due to limited ABI specification coverage, all the native linker implementations limit the function
```
```suggestion
 * All the native linker implementations support function function descriptors that contain only so-called ...
```

JornVernee replied:
So, just drop the 'Due to limited ABI specification coverage'? I guess it's not really needed.

JornVernee commented on src/java.base/share/classes/java/lang/foreign/Linker.java:
```diff
@@ -195,6 +196,18 @@
+ * <p>
+ * Due to limited ABI specification coverage, all the native linker implementations limit the function
+ * descriptors that they support to those that contain only so-called <em>canonical</em> layouts. These layouts
+ * have the following restrictions:
```
The third item in the list is for group size needing to be a multiple of the alignment constraint, the forth item refers to that, so I'll keep the ordered list.

WRT byte order: I think it's fine to just say "native byte order". Keep it simple. Or do you think there might be confusion about the meaning of "native"? (It looks like not even ByteOrder::nativeOrder explains that term: https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/nio/ByteOrder.html#nativeOrder())

"padding layout elements" is not quite right I think, since they also have to be the right size. I think saying just "padding" is more accurate (and simpler).

I'll switch to using plain links instead.

JornVernee replied:
Uploaded another version

JornVernee commented on src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:
```diff
@@ -101,22 +107,63 @@ public SystemLookup defaultLookup() {
-        if (layout instanceof GroupLayout gl) {
-            for (MemoryLayout member : gl.memberLayouts()) {
-                checkHasNaturalAlignmentRecursive(member);
+        checkByteOrder(layout);
```
Yes, will do.

mcimadamore reviewed (approved):
Looks good.

One (optional, for future?) thing that came to mind, is that if we wanted to speed up the check we could compute whether a layout is canonical or not on construction (since for structs we already have to go through all the elements and check alignment).

JornVernee commented:
@mcimadamore Thanks for the review!

I've created a CSR with the finalized javadoc here: https://bugs.openjdk.org/browse/JDK-8306631 Could you review that as well? TIA

JornVernee commented:
/integrate
